oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtEncoder.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.jwt;
+
+/**
+ * Implementations of this interface are responsible for encoding a JSON Web Token (JWT)
+ * to it's compact claims representation format.
+ *
+ * <p>
+ * JWTs may be represented using the JWS Compact Serialization format for a JSON Web
+ * Signature (JWS) structure or JWE Compact Serialization format for a JSON Web Encryption
+ * (JWE) structure. Therefore, implementors are responsible for signing a JWS and/or
+ * encrypting a JWE.
+ *
+ * @author Anoop Garlapati
+ * @author Joe Grandja
+ * @since 5.5
+ * @see Jwt
+ * @see JoseHeader
+ * @see JwtClaimsSet
+ * @see JwtDecoder
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7519">JSON Web Token
+ * (JWT)</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7515">JSON Web Signature
+ * (JWS)</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7516">JSON Web Encryption
+ * (JWE)</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7515#section-3.1">JWS
+ * Compact Serialization</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7516#section-3.1">JWE
+ * Compact Serialization</a>
+ */
+@FunctionalInterface
+public interface JwtEncoder {
+
+	/**
+	 * Encode the JWT to it's compact claims representation format.
+	 * @param headers the JOSE header
+	 * @param claims the JWT Claims Set
+	 * @return a {@link Jwt}
+	 * @throws JwtEncodingException if an error occurs while attempting to encode the JWT
+	 */
+	Jwt encode(JoseHeader headers, JwtClaimsSet claims) throws JwtEncodingException;
```

Comment:
I'm in favour of keeping the operation as-is or consider changing it to:

`Jwt encode(JoseHeader.Builder headersBuilder, JwtClaimsSet.Builder claimsBuilder)`

Introducing the [JoseHeader](https://tools.ietf.org/html/rfc7519#section-5) and [JwtClaimsSet](https://tools.ietf.org/html/rfc7519#section-4) constructs aligns with the spec.

The `Jwt.Builder` currently supports setting the typed and custom claims but the headers only generically via `header()` or `headers()`. Adding all the possible JWS and JWE headers to `Builder` is going to make the `Builder` quite large and not as easy to understand without the clear separation of claims and headers, which the new constructs would achieve.

I would even prefer to re-work `Jwt`, specifically the `Builder` and associated factory method `withTokenValue(String tokenValue)`. For example, some of the proposed changes would be:

```java
`JoseHeader Jwt.getHeaders()`

`JwtClaimsSet Jwt.getClaims()`

static Jwt.with(JoseHeader, JwtClaimsSet)
```

I know what I'm proposing as far as the re-work for `Jwt` is a big change but I strongly feel we should have separate representations of the JOSE headers and Jwt Claims set. I believe this will keep things simpler (in each class), easier to maintain and easier to grow when we introduce support for JWE at a later point (which adds new headers to the mix).

@jzheaux @rwinch What are your thoughts on this?

Reply:
> For the Nested JWT use case, the returned object cannot be a `Jwt`.


I don't really see why this is the case, could you elaborate? Couldn't `Jwt` contain three members, like so:

```java
JwtClaimsSet getClaimsSet();
JwsHeader getJwsHeader();
JweHeader getJweHeader();
```

Flattening a structure into its relevant parts like this is a practical way for an application to consume it. It's uncommon for folks to need to know the headers anyway -- most callers only need to know the resulting encoded value and its claims.

I imagine that a change like this will have to happen anyway to `Jwt` so that `Jwt NimbusJwtDecoder#decode(String)` can decrypt and verify an encrypted and signed JWT.

If `Jwt` cannot be made to suit, I wonder if this means that `JwtEncoder#encode`'s return type needs to be different from `Jwt`. I believe returning a `String` can address this concern.

Practically speaking, returning a serialized value is less of a risk than it may seem on the surface. In all cases, the claims must be specified by the caller, so the caller does not lose access to the claims when a string is returned. I believe it will be uncommon for the caller to need to know the headers that were used -- but if this is needed, an application can easily parse the return value with something like `JWTParser.parse` from Nimbus.

The use case we currently have in the codebase, `NimbusJwtClientAuthenticationParametersConverter`, only needs the serialized value. A brief Find Usages search in Spring Authorization Server shows that only the serialized value and the claims are needed to fulfill those use cases.

If `Jwt` cannot be made to suit and if `String` feels too risky, I wonder if we could introduce `EncodedJwt` and return that. It could initially contain only the encoded `String`. Additional members could be added if needed.

I'm keen on making this aspect of the contract work. Given that `JwtDecoder` has the potential to decode both signed JWTs and Nested JWTs because of its contract, I think it's reasonable to see if `JwtEncoder` can encode both signed JWTs and Nested JWTs with the right contract. Similar to applications asking Spring Security to "please decrypt and verify this" without exposing them to Nested JWT internals (Nimbus does exactly this), it seems quite reasonable to give Spring Security a set of claims and say "please sign and encrypt this". 

Parenthetically, if we can't do this, I think it will be less confusing if the interface is called `JwsEncoder` so that users don't think to align the capabilities of `JwtDecoder` with `JwtEncoder`.

Reply:
I'm marking this as resolved because the code now uses JwtEncoderParameters. We can add any additional feedback on the updated code.